### PR TITLE
go proto: include (opt-in) vtprotobuf generation

### PR DESCRIPTION
**Pull request by howardjohn**
Commit Message: go proto: include vtprotobuf generation
Additional Description:

See https://github.com/envoyproxy/go-control-plane/issues/824 for more information

This PR adds the vtprotobuf protoc plugin for Go. This works _on top_ of the existing protoc-gen-go, to add optimized marshal functions that callers can _opt in_ to using. This is not like `gogo`, which was a very invasive change -- everything is layered and opt-in. See issue for benchmarks, etc.

Additionally, to avoid possible binary size increase, the entire new code is protected under a go build tag. Users will need to opt-in at build time (`-tags=vtprotobuf`). **By default, there is no impact for users at all**.

Risk Level: Low - only additional opt-in code
Testing: Manually tested in Istio codebase
Docs Changes: NONE
Release Notes: NONE
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: bazel/dependency_imports.bzl
Comment (moderation) [inline bazel/dependency_imports.bzl:151]: OSSF Scorecard results:
```
scorecard --repo=https://github.com/planetscale/vtprotobuf
RESULTS
-------
Aggregate score: 4.2 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                                               DOCUMENTATION/REMEDIATION                                               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Branch-Protection      | branch protection not enabled  | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#branch-protection      |
|         |                        | on development/release         |                                                                                                                       |
|         |                        | branches                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | CI-Tests               | 6 out of 6 merged PRs          | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#ci-tests               |
|         |                        | checked by a CI test -- score  |                                                                                                                       |
|         |                        | normalized to 10               |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no effort to earn an OpenSSF   | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#cii-best-practices     |
|         |                        | best practices badge detected  |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 7 / 10  | Code-Review            | found 2 unreviewed changesets  | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#code-review            |
|         |                        | out of 7 -- score normalized   |                                                                                                                       |
|         |                        | to 7                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Contributors           | project has 13 contributing    | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#contributors           |
|         |                        | companies or organizations     |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Dangerous-Workflow     | no dangerous workflow patterns | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#dangerous-workflow     |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Dependency-Update-Tool | no update tool detected        | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed          | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | License                | license file detected          | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Maintained             | 30 commit(s) and 7 issue       | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#maintained             |
|         |                        | activity found in the last 90  |                                                                                                                       |
|         |                        | days -- score normalized to 10 |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Packaging              | packaging workflow not         | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#packaging              |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 2 / 10  | Pinned-Dependencies    | dependency not pinned by hash  | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#pinned-dependencies    |
|         |                        | detected -- score normalized   |                                                                                                                       |
|         |                        | to 2                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | SAST tool is not run on all    | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#sast                   |
|         |                        | commits -- score normalized to |                                                                                                                       |
|         |                        | 0                              |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Security-Policy        | security policy file not       | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#security-policy        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Signed-Releases        | no releases found              | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#signed-releases        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Token-Permissions      | detected GitHub workflow       | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#token-permissions      |
|         |                        | tokens with excessive          |                                                                                                                       |
|         |                        | permissions                    |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Vulnerabilities        | 14 existing vulnerabilities    | https://github.com/ossf/scorecard/blob/483cc31b60c63cc7e13a9b4211ff6acf8b4d7d20/docs/checks.md#vulnerabilities        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
```

**Review by moderation — COMMENTED** [review]
For the binary size risk you highlight, is that risk for the `go-control-plane` binary, the Envoy binary or both?

> Binary size: importing every proto results in a 4MB increase in binary size. Not its quite uncommon to import every proto in the repo, so typically this cost is much smaller
Mitigation, if needed: expose this only under a build tag, so a use can opt in (or opt out)

**Comment by howardjohn** [comment]
> For the binary size risk you highlight, is that risk for the `go-control-plane` binary, the Envoy binary or both?
> 
> > Binary size: importing every proto results in a 4MB increase in binary size. Not its quite uncommon to import every proto in the repo, so typically this cost is much smaller
> > Mitigation, if needed: expose this only under a build tag, so a use can opt in (or opt out)

this exclusively impacts go-control-plane (binary and importers of the library)

**Comment by kyessenov** [comment]
cc @dfawley since grpc-go consumes go-control-plane

File: bazel/dependency_imports.bzl
Comment (moderation) [reply bazel/dependency_imports.bzl:151]: I'm not seeing those changes. I'm not seeing branch protection and your run doesn't see CI. Adding a security policy would be good too.
```
scorecard version
         __  ____     ____    ___    ____    _____    ____      _      ____    ____
        / / / ___|   / ___|  / _ \  |  _ \  | ____|  / ___|    / \    |  _ \  |  _ \
       / /  \___ \  | |     | | | | | |_) | |  _|   | |       / _ \   | |_) | | | | |
  _   / /    ___) | | |___  | |_| | |  _ <  | |___  | |___   / ___ \  |  _ <  | |_| |
 (_) /_/    |____/   \____|  \___/  |_| \_\ |_____|  \____| /_/   \_\ |_| \_\ |____/
./scorecard: OpenSSF Scorecard

GitVersion:    v4.13.1-84-gcb721a85-dirty
GitCommit:     cb721a8526fccd47bc234d9d2005eecd01e4b0ae
GitTreeState:  dirty
BuildDate:     1701764656
GoVersion:     go1.21.4
Compiler:      gc
Platform:      linux/amd64
```
```
scorecard --repo=https://github.com/planetscale/vtprotobuf
RESULTS
-------
Aggregate score: 4.9 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                                               DOCUMENTATION/REMEDIATION                                               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Branch-Protection      | branch protection not enabled  | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#branch-protection      |
|         |                        | on development/release         |                                                                                                                       |
|         |                        | branches                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | CI-Tests               | 6 out of 6 merged PRs          | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#ci-tests               |
|         |                        | checked by a CI test -- score  |                                                                                                                       |
|         |                        | normalized to 10               |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no effort to earn an OpenSSF   | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#cii-best-practices     |
|         |                        | best practices badge detected  |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 8 / 10  | Code-Review            | found 1 unreviewed changesets  | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#code-review            |
|         |                        | out of 7 -- score normalized   |                                                                                                                       |
|         |                        | to 8                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Contributors           | project has 13 contributing    | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#contributors           |
|         |                        | companies or organizations     |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Dangerous-Workflow     | no dangerous workflow patterns | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#dangerous-workflow     |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Dependency-Update-Tool | no update tool detected        | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed          | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | License                | license file detected          | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Maintained             | 30 commit(s) and 6 issue       | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#maintained             |
|         |                        | activity found in the last 90  |                                                                                                                       |
|         |                        | days -- score normalized to 10 |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Packaging              | packaging workflow not         | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#packaging              |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 2 / 10  | Pinned-Dependencies    | dependency not pinned by hash  | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#pinned-dependencies    |
|         |                        | detected -- score normalized   |                                                                                                                       |
|         |                        | to 2                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | SAST tool is not run on all    | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#sast                   |
|         |                        | commits -- score normalized to |                                                                                                                       |
|         |                        | 0                              |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Security-Policy        | security policy file not       | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#security-policy        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Signed-Releases        | no releases found              | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#signed-releases        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Token-Permissions      | detected GitHub workflow       | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#token-permissions      |
|         |                        | tokens with excessive          |                                                                                                                       |
|         |                        | permissions                    |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 8 / 10  | Vulnerabilities        | 2 existing vulnerabilities     | https://github.com/ossf/scorecard/blob/cb721a8526fccd47bc234d9d2005eecd01e4b0ae/docs/checks.md#vulnerabilities        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
```

Comment (howardjohn) [reply bazel/dependency_imports.bzl:151]: Vulnerabilities went from 0 (your original) -> 10 (my run) -> 8 (your run).
Looks like I was on an older version of scorecard, hence the differences

**Comment by dfawley** [comment]
> cc @dfawley since grpc-go consumes go-control-plane

Thank you.  IIUC this should not affect us since we don't use bazel?

**Comment by howardjohn** [comment]
> > cc @dfawley since grpc-go consumes go-control-plane
> 
> Thank you. IIUC this should not affect us since we don't use bazel?

This will impact the generated code in go-control-plane (everything under https://github.com/envoyproxy/go-control-plane/tree/main/envoy)

**Comment by dfawley** [comment]
I see... 

> Binary size: importing every proto results in a 4MB increase in binary size

How can I see what the impact will be to grpc-go binary sizes with xds enabled?

Our set of protos imported is currently:
```go
"github.com/envoyproxy/go-control-plane/envoy/admin/v3"
"github.com/envoyproxy/go-control-plane/envoy/api/v2"
"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
"github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
"github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
"github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
"github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
"github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
"github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2"
"github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
"github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/clusters/aggregate/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/filters/common/fault/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/fault/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/rbac/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/router/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/client_side_weighted_round_robin/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/least_request/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/pick_first/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/ring_hash/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/round_robin/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/wrr_locality/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/rbac/audit_loggers/stream/v3"
"github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
"github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
"github.com/envoyproxy/go-control-plane/envoy/service/load_stats/v3"
"github.com/envoyproxy/go-control-plane/envoy/service/status/v3"
"github.com/envoyproxy/go-control-plane/envoy/type"
"github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
"github.com/envoyproxy/go-control-plane/envoy/type/v3"
"github.com/envoyproxy/go-control-plane/pkg/cache/types"
"github.com/envoyproxy/go-control-plane/pkg/cache/v3"
"github.com/envoyproxy/go-control-plane/pkg/resource/v3"
"github.com/envoyproxy/go-control-plane/pkg/server/v3"
"github.com/envoyproxy/go-control-plane/pkg/wellknown"
```

**Comment by howardjohn** [comment]
> I see...
> 
> > Binary size: importing every proto results in a 4MB increase in binary size
> 
> How can I see what the impact will be to grpc-go binary sizes with xds enabled?
> 
> Our set of protos imported is currently:
> 
> ```go
> "github.com/envoyproxy/go-control-plane/envoy/admin/v3"
> "github.com/envoyproxy/go-control-plane/envoy/api/v2"
> "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
> "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
> "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
> "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
> "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
> "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
> "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2"
> "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
> "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/clusters/aggregate/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/common/fault/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/fault/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/rbac/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/router/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/client_side_weighted_round_robin/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/least_request/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/pick_first/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/ring_hash/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/round_robin/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/wrr_locality/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/rbac/audit_loggers/stream/v3"
> "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
> "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
> "github.com/envoyproxy/go-control-plane/envoy/service/load_stats/v3"
> "github.com/envoyproxy/go-control-plane/envoy/service/status/v3"
> "github.com/envoyproxy/go-control-plane/envoy/type"
> "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
> "github.com/envoyproxy/go-control-plane/envoy/type/v3"
> "github.com/envoyproxy/go-control-plane/pkg/cache/types"
> "github.com/envoyproxy/go-control-plane/pkg/cache/v3"
> "github.com/envoyproxy/go-control-plane/pkg/resource/v3"
> "github.com/envoyproxy/go-control-plane/pkg/server/v3"
> "github.com/envoyproxy/go-control-plane/pkg/wellknown"
> ```

Using `examples/features/xds/client` and `go build -ldflags '-extldflags -static -s -w' .`:

Before: 21M (21368832 bytes)
After: 23M (24104960 bytes)

`go mod edit -replace github.com/envoyproxy/go-control-plane=github.com/howardjohn/go-control-plane@v0.0.0-20231204221533-4e477fbf1a5b` can temporarily test the change before this merges

**Comment by soulxu** [comment]
@adisuissa Would you like to take a look at this? but let me know if I'm assigned to the wrong person. Thanks!

**Comment by adisuissa** [comment]
Thanks for working on this!
Generally speaking I'm ok with this assuming the new dependency is ok, but want to highlight the risks that you [mentioned](https://github.com/envoyproxy/go-control-plane/issues/824#issue-2007122965):
> Binary size: importing every proto results in a 4MB increase in binary size. Not its quite uncommon to import every proto in the repo, so typically this cost is much smaller
>       Mitigation, if needed: expose this only under a build tag, so a use can opt in (or opt out)
> Increase dependency that could, in theory, start to decay.
>       Mitigation: dropping vtprotobuf would be compatible, as noted above (just have a large performance impact)

Specifically the binary size increase may be impacting Envoy-Mobile. @alyssawilk to chime in on this.

**Comment by howardjohn** [comment]
> Specifically the binary size increase may be impacting Envoy-Mobile. @alyssawilk to chime in on this.
The impact is only for golang code which I assume doesn't impact mobile (?)

**Comment by howardjohn** [comment]
It seems like a lot of hangup is on binary size. What if we make it opt-in behind a go build tag?

**Comment by abeyad** [comment]
The binary size will matter a lot for Envoy Mobile, 4MB is quite a bit to add to the binary size of a mobile app library.  It would be interesting to see what the Envoy Mobile binary size impact is with and without this PR.

@howardjohn could you kindly run the following commands with and without your branch, and report back on the results:
1. `bazel build --config=remote //test/performance:test_binary_size --config=sizeopt --copt=-ggdb3 --linkopt=-fuse-ld=lld --config=release-android`
2. `strip -s -o bazel-bin/test/performance/test_binary_size_stripped bazel-bin/test/performance/test_binary_size`
3. `ls -alh bazel-bin/test/performance/test_binary_size_stripped`

**Comment by howardjohn** [comment]
> The binary size will matter a lot for Envoy Mobile, 4MB is quite a bit to add to the binary size of a mobile app library. It would be interesting to see what the Envoy Mobile binary size impact is with and without this PR.
> 
> @howardjohn could you kindly run the following commands with and without your branch, and report back on the results:
> 
> 1. `bazel build --config=remote //test/performance:test_binary_size --config=sizeopt --copt=-ggdb3 --linkopt=-fuse-ld=lld --config=release-android`
> 2. `strip -s -o bazel-bin/test/performance/test_binary_size_stripped bazel-bin/test/performance/test_binary_size`
> 3. `ls -alh bazel-bin/test/performance/test_binary_size_stripped`

This only impacts Go (via go-control-plane) so its not relevant to Envoy mobile

**Comment by abeyad** [comment]
> > The binary size will matter a lot for Envoy Mobile, 4MB is quite a bit to add to the binary size of a mobile app library. It would be interesting to see what the Envoy Mobile binary size impact is with and without this PR.
> > @howardjohn could you kindly run the following commands with and without your branch, and report back on the results:
> > 
> > 1. `bazel build --config=remote //test/performance:test_binary_size --config=sizeopt --copt=-ggdb3 --linkopt=-fuse-ld=lld --config=release-android`
> > 2. `strip -s -o bazel-bin/test/performance/test_binary_size_stripped bazel-bin/test/performance/test_binary_size`
> > 3. `ls -alh bazel-bin/test/performance/test_binary_size_stripped`
> 
> This only impacts Go (via go-control-plane) so its not relevant to Envoy mobile

So it wouldn't get compiled into the Envoy Mobile .so?  Based on your PR, it looks like it effects the api_proto_package which EM would indirectly use.

**Comment by howardjohn** [comment]
> So it wouldn't get compiled into the Envoy Mobile .so? Based on your PR, it looks like it effects the api_proto_package which EM would indirectly use.

I don't see how it would, it only impacts go-control-plane. I cannot run the commands you sent to verify though, they do not seem to work for me.

**Comment by abeyad** [comment]
> I don't see how it would, it only impacts go-control-plane. I cannot run the commands you sent to verify though, they do not seem to work for me.

I tried it and the binary size didn't change for Envoy Mobile.  So I don't have concerns with this PR from the Envoy Mobile point of view.  Thanks!

**Review by adisuissa — COMMENTED** [review]
Given @abeyad ok, this PR LGTM from change perspective.
@moderation to weigh in on dependency perspective.

**Comment by dfawley** [comment]
I'm not sure this is a good idea.

- This can never be removed once it is included.
- It has downsides (binary size) to go along with its benefits (performance).
- It might prevent us from upgrading protoc-gen-go if there is ever an incompatible change made in protoc-gen-go and this project stops being maintained.

Can vtproto not work as a separate add-on?  Why does it need to generate code into the same package as the protoc-gen-go code?

**Comment by howardjohn** [comment]
> * This can never be removed once it is included.

I disagree. vtprotobuf is designed in a way such that the same program can compile with or without it included (https://github.com/vitessio/vitess/blob/40f314c1d052db3fa5a52e5fb2ddd2bddaefde44/go/vt/servenv/grpc_codec.go#L40-L49 for example). This gives envoy flexibility to rip it out at anytime without breaking compatibility with importers (of course, they don't get the benefits anymore). This is very different from gogo proto.

> * It has downsides (binary size) to go along with its benefits (performance).

See https://github.com/envoyproxy/envoy/pull/31172#issuecomment-1852199211 and https://github.com/planetscale/vtprotobuf/pull/122, I intend to change this to make it opt-in.

> * It might prevent us from upgrading protoc-gen-go if there is ever an incompatible change made in protoc-gen-go and this project stops being maintained.

If so, we can remove it per above. I would not expect protoc-gen-go to make such changes given how things are layered; any change that could break vtprotobuf would laso likely completely break compatibility and represent a protobuf-go v3. vtproto doesn't depend on any internal details of protoc-gen-go really, aside from the fact a Go struct exists.

> Can vtproto not work as a separate add-on? Why does it need to generate code into the same package as the protoc-gen-go code?

It is attaching methods to the types which can only be done in the same package

**Comment by dfawley** [comment]
> It is attaching methods to the types which can only be done in the same package

Is that really the only way they could find to improve the performance?

Or are there maybe other options that would allow this project to benefit everyone without requiring someone to go modify the official codegen for every project that has protos?

**Comment by howardjohn** [comment]
With the most recent commit this is now **opt-in**. Any user that doesn't explicitly opt in should have no impact.

**Comment by phlax** [comment]
/wait-any for further review - cc @moderation @adisuissa

**Comment by phlax** [comment]
@howardjohn seems DCO is broken

**Comment by howardjohn** [comment]
> @howardjohn seems DCO is broken

Thanks, fixed.

**Comment by adisuissa** [comment]
cc @dfawley and @htuch for any comments, as they've probably had more experience in this domain.

**Comment by dfawley** [comment]
If this is opt-in, I have no concerns from the gRPC side, as our users will be unaffected (unless they choose to enable it in their builds).

**Comment by adisuissa** [comment]
Please merge main.
Change LGTM. Assigning @htuch for senior-maintainer review.
/assign @htuch

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Comment by howardjohn** [comment]
@adisuissa thanks! I merged. Test is failing due to the v1.29.0 publishing I think

**Review by htuch — APPROVED** [review]
Makes sense as opt-in improvement, thanks.

**Pull request merged**
